Rubycritic compare b/w branches

HemanthMudalaiah opened the pull request.

Commit:
This commit will have changes in rubycritic where you can compare two branches

Commit:
This commit will have changes in rubycritic where you can compare two branches

Commit:
This commit has changes where we can decide which of the file changes degraded the code quality

Commit:
This commit has changes where we can decide which of the file changes degraded the code quality

Commit:
Comparing the diff with the whole branch for duplication analysis

Commit:
Comparing the diff with the whole branch for duplication analysis

Commit:
Configuring smell link to point to feature branch code

Commit:
Configuring smell link to point to feature branch code

Commit:
Changes in the content of the build details

Commit:
Changes in the content of the build details

Commit:
Gitlab Integration

Pushing the status of the code and report location to gitlab using gitlab api

Commit:
Gitlab Integration

Pushing the status of the code and report location to gitlab using gitlab api

Commit:
Code cleanup

* Code refactoring, DRY
* Comparing between branches in local without pr id.

Commit:
Code cleanup

* Code refactoring, DRY
* Comparing between branches in local without pr id.

rohitcy committed:
Revamp the web UI

Commit:
Specs and code cleanup

1. Setting a new mode to compare between branches.
2. Moved the code changes to a seperate file.
3. Pushing comments on to pr.
4. Specs for analysed_module_collection.

Commit:
Specs and code cleanup

1. Setting a new mode to compare between branches.
2. Moved the code changes to a seperate file.
3. Pushing comments on to pr.
4. Specs for analysed_module_collection.

tejasbubane committed:
Fix some ruby warnings thrown on rake test

Upgrade `parser` to `2.3.1.4`.
One warning still left which comes from `parser`. Hopefully that is
fixed in later versions.

troessner committed:
Merge pull request #183 from tejasbubane/fix-warnings

Fix some ruby warnings thrown on rake test

Commit:
Code Cleanup & Refactoring

1. Removed Gitlab,Github integrations.
2. Template Changes.
3. Options changes.

Commit:
Code Cleanup & Refactoring

1. Removed Gitlab,Github integrations.
2. Template Changes.
3. Options changes.

Commit:
Added specs and threshold option

1. Added specs for compare module
2. Added threshold as an option in rubycritic before it was as app side configuration
3. Check for any uncommited changes

Commit:
Added specs and threshold option

1. Added specs for compare module
2. Added threshold as an option in rubycritic before it was as app side configuration
3. Check for any uncommited changes

Commit:
Removing simplecov from dependency

Commit:
Refactoring compare module

1. Removed redundant code.
2. Adding config so that browser doesn't open html report during test.

Commit:
Refactoring compare module

1. Removed redundant code.
2. Adding config so that browser doesn't open html report during test.

tejasbubane commented on lib/rubycritic/cli/options.rb:
```diff
@@ -19,6 +19,16 @@ def parse
 
+          opts.on('-b', '--branch BASE_BRANCH,FEATURE_BRANCH' , 'Set branches') do |branches|
+            self.base_branch = String(branches.split(',')[0].strip)
+            self.feature_branch = String(branches.split(',')[1].strip)
```
You can use multiple assignment here and use `.to_s` instead of `String()`

tejasbubane replied:
``` ruby
self.base_branch, self.feature_branch = branches.split(',').map { |b| String(b.strip) }
```

tejasbubane commented on lib/rubycritic/cli/options.rb:
```diff
@@ -71,14 +81,17 @@ def to_h
 
       attr_accessor :mode, :root, :format, :deduplicate_symlinks,
-                    :suppress_ratings, :minimum_score, :no_browser, :parser
+                    :suppress_ratings, :minimum_score, :no_browser, :parser, :base_branch, :feature_branch, :threshold_score
```
line too long

tejasbubane commented on lib/rubycritic/commands/compare.rb:
```diff
@@ -0,0 +1,136 @@
+        @files_affected = []
+        @analysed_modules = []
+        @number = 0
+        @code_index_file_location = ''
```
are these all variables used? If not remove them. Also give better names - eg `number` is not very intuitive. Also I don't prefer names with data structures like `hash`, `array`, etc.

tejasbubane commented on lib/rubycritic/commands/compare.rb:
```diff
@@ -0,0 +1,136 @@
+      end
+
+      def update_build_number
+        build_file_location = '/tmp/build_count.txt'
```
do we need this?

tejasbubane replied:
Mention the use of this count file as a comment above the method.

tejasbubane commented on lib/rubycritic/commands/compare.rb:
```diff
@@ -0,0 +1,136 @@
+      end
+
+      def set_scores(branch_type, score)
+        branch_type == 'base_branch' ? Config.base_branch_score = score : Config.feature_branch_score = score
```
``` ruby
Config.send(:"#{branch_type}_score") = score
```

tejasbubane commented on lib/rubycritic/commands/compare.rb:
```diff
@@ -0,0 +1,136 @@
+      end
+
+      def threshold_values_set?
+        Config.threshold_score > 0
```
if `threshold` is `nil` this will fail with `undefined method`>' for nil:NilClass`

tejasbubane replied:
Fix this

tejasbubane commented on test/lib/rubycritic/commands/compare_test.rb:
```diff
@@ -0,0 +1,88 @@
+          true
+        elsif arg == 'feature_branch'
+          true
+        end
```
What is this? 😳

Commit:
Code cleanup

1. Changes for stub in compare test file
2. Cleaning up compare.rb
3. Changes in option parsing

Commit:
Code cleanup

1. Changes for stub in compare test file
2. Cleaning up compare.rb
3. Changes in option parsing

tejasbubane commented on features/command_line_interface/options.feature:
```diff
@@ -21,6 +21,10 @@ Feature: RubyCritic can be controlled using command-line options
           -p, --path [PATH]                Set path where report will be saved (tmp/rubycritic by default)
+          -b BASE_BRANCH,FEATURE_BRANCH,   Set branches
+              --branch
+          -t [THRESHOLD_SCORE],            Set a threshold score
```
mention that this works only with `-b`

tejasbubane commented on lib/rubycritic/commands/compare.rb:
```diff
@@ -0,0 +1,130 @@
+
+      def degraded_files
+        files_affected = []
+        @feature_branch.each do |key, value|
```
use `map` instead of `each`

tejasbubane commented on test/lib/rubycritic/core/analysed_modules_collection_test.rb:
```diff
@@ -38,6 +38,40 @@
+      let(:analysed_modules) { [RubyCritic::AnalysedModule.new(pathname: Pathname.new('test/samples/empty.rb'), name: 'Empty'),
+                                RubyCritic::AnalysedModule.new(pathname: Pathname.new('test/samples/unparsable.rb'), name: 'Unparsable'),
+                                ] 
+                              }
```
use `do-end` for multi-line blocks.

Commit:
Options and compare changes

Commit:
Options and compare changes

nunosilva800 committed:
Merge pull request #184 from cybrilla/new-web-ui

New Web UI

nunosilva800 committed:
Bump to v3.1.0

troessner committed:
Merge pull request #185 from whitesmith/bump-to-3.1

Bump to v3.1.0

nav16 committed:
Add screenshot of new ui [ci skip]

troessner committed:
Merge pull request #186 from nav16/update-readme

Add screenshots of new ui in README

vijayanandnandam commented:
@HemanthMudalaiah get this done quickly and raise PR on `rubycritic`

Commit:
Fixing the options fetaure spec

Commit:
Fixing the options fetaure spec

tejasbubane commented on lib/rubycritic/commands/compare.rb:
```diff
@@ -0,0 +1,129 @@
+
+      def build_details
+        details = "Base branch (#{Config.base_branch}) score: " + Config.base_branch_score.to_s + "\n"
+        details += "Feature branch (#{Config.feature_branch}) score: " + Config.feature_branch_score.to_s + "\n"
```
use multi-line string

tejasbubane commented on lib/rubycritic/commands/compare.rb:
```diff
@@ -0,0 +1,129 @@
+      end
+
+      def degraded_files
+        files_affected = []
```
remove this, return array from map

tejasbubane commented on lib/rubycritic/commands/compare.rb:
```diff
@@ -0,0 +1,129 @@
+      end
+
+      def build_cost_hash(cost_hash, analysed_modules)
+        complexity_hash = eval("@#{cost_hash}")
```
remove `eval`

tejasbubane commented on lib/rubycritic/commands/compare.rb:
```diff
@@ -0,0 +1,129 @@
+
+      def build_cost_hash(cost_hash, analysed_modules)
+        complexity_hash = eval("@#{cost_hash}")
+        analysed_modules.map { |analysed_module| complexity_hash.merge!(:"#{analysed_module.name}" => analysed_module.cost) }
```
line is too long

tejasbubane commented on test/lib/rubycritic/commands/compare_test.rb:
```diff
@@ -0,0 +1,49 @@
+    class Git < Base
+      def self.switch_branch(branch)
+        File.open('test/samples/compare_file.rb', 'w') { |file| file.truncate(0) }
+        File.open('test/samples/compare_file.rb', 'w') { |file| file.puts File.readlines("test/samples/#{branch}_file.rb") }
```
line too long

Commit:
Code cleanup and changes

1. Generating report for all modified files in the feature branch instead of degraded files as before.
2. Removing methods related to get degraded files.
3. Fixing specs due the above changse.

Commit:
Code cleanup and changes

1. Generating report for all modified files in the feature branch instead of degraded files as before.
2. Removing methods related to get degraded files.
3. Fixing specs due the above changse.

Commit:
Spec fix

Commit:
Spec fix

tejasbubane commented on lib/rubycritic/commands/compare.rb:
```diff
@@ -0,0 +1,125 @@
+      def execute
+        Config.no_browser = true
+        compare_branches
+        status_reporter.score = (Config.base_branch_score - Config.feature_branch_score).abs
```
tell whether score has increased or decreased in the console output.

tejasbubane commented on lib/rubycritic/commands/compare.rb:
```diff
@@ -0,0 +1,125 @@
+      # create a txt file with the branch score details
+      def build_details
+        details = "Base branch (#{Config.base_branch}) score: " + Config.base_branch_score.to_s + "\n"\
+                  "Feature branch (#{Config.feature_branch}) score: " + Config.feature_branch_score.to_s + "\n"
```
use string interpolation

Commit:
Abort message changes

Commit:
Abort message changes

Commit:
Adding links in the navbar

Commit:
Merge Fixes

tejasbubane commented:
Something went wrong with the rebase. It contains commits from others as well.
